data/campaigns/fri05.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,229 @@
+      sleep(7200)
+   end
+end
+
+function check_objective_wood()
+   local o = add_campaign_objective(obj_wood)
+   while not check_for_buildings(p1, {frisians_woodcutters_house = 3, frisians_foresters_house = 4}) do
+      sleep(4273)
+   end
+   set_objective_done(o)
+   campaign_message_box(secured_wood)
+end
+
+function check_objective_block()
+   local o = add_campaign_objective(obj_block)
+   while true do
+      sleep(4916)
+      local sites = 0
+      for x = 84, 94 do
+         for y = 114, 122 do
+            local f = map:get_field(x, y)
+            if f.immovable ~= nil and f.immovable.descr.type_name == "militarysite" and f.immovable.fields[1] == f and f.owner == p1 then
+               sites = sites + 1
+               if sites >= 3 then
+                  set_objective_done(o)
+                  campaign_message_box(secured_block)
+                  return
+               end
+            end
+         end
+      end
+   end
+end
+
+function check_objective_uplands()
+   local o = add_campaign_objective(obj_uplands)
+   while true do
+      sleep(4471)
+      for _,wh in ipairs(p1:get_buildings("frisians_warehouse")) do
+         if wh.fields[1].height > 40 then
+            set_objective_done(o)
+            campaign_message_box(secured_uplands)
+            return
+         end
+      end
+   end
+end
+
+function mission_thread()
+   include "map:scripting/init_p2.lua"
+   run(check_fish)
+   campaign_message_box(intro_1)
+   campaign_message_box(intro_2)
+   campaign_message_box(intro_3)
+   campaign_message_box(intro_4)
+
+   scroll_to_field(map.player_slots[1].starting_field)
+   include "map:scripting/init_p1.lua"
+   p1:reveal_fields(map.player_slots[2].starting_field:region(9))
+   local reveal_fields = {}
+   for x=0,121 do
+      for y=17,240 do
+         local f = map:get_field(x, y)
+         p2:reveal_fields({f})
+         if p1:sees_field(f) then
+            p1:reveal_fields({f})  -- Make currently seen fields permanently visible without hiding them first
+         else
+            table.insert(reveal_fields, f)
+         end
+      end
+   end
+   run(function()
+      reveal_randomly(p1, reveal_fields, 100)
+      reveal_fields = nil
+   end)
+   sleep(5000)
+
+   campaign_message_box(intro_5)
+   campaign_message_box(intro_6)
+   campaign_message_box(intro_7)
+
+   scroll_to_field(map.player_slots[2].starting_field)
+   sleep(10000)
+   scroll_to_field(map.player_slots[1].starting_field)
+
+   campaign_message_box(intro_8)
+   campaign_message_box(intro_9)
+   campaign_message_box(intro_10)
+   local o = add_campaign_objective(obj_fight)
+   sleep(1500)
+
+   campaign_message_box(trade_1)
+   campaign_message_box(trade_2)
+
+   for i,f in ipairs({
+      map:get_field(75, 107),
+      map:get_field(55,  75),
+      map:get_field(33,  72),
+      map:get_field(65, 117),
+      map:get_field(79, 161),
+      map:get_field(89, 177),
+   }) do
+      sleep(750)
+      scroll_to_field(f)
+   end
+
+   sleep(2500)
```

Comment:
in the following the player gets walls of text that are separated by annoying intervals. each time he wants to start doing something he is interrupted again by a message box. So for the intermediate sleeps please consider reducing them a lot.
furthermore Im believe wecould safely distribute the infos a bit better over the storyline evolving
e.g. 
1. objectives Uplands, block and wood shoud be the first ones. 
2. objective trade could be given when one of the traded wares gets low most probably coal.   
3. hint about diking coulod be given if we have more then e.g. 50 log on stock
4. text about training (perhaps together with a hint to upgrade /built new armor smithies) can be given after we have made e.g. 10 succesful trades.

Reply:
do you want to do anything about this? 
Shall I prepare a PR with a suggestion?

Reply:
Those are good suggestions but I'm having a bit of "writer's block" currently about rephrasing the texts. So if you want to make a PR for this it would be appreciated :)